orangecontrib/text/widgets/owtopicmodeling.py

```diff
@@ -0,0 +1,310 @@
+            self.widgets.append(widget)
+            setattr(self, attr_name, widget)
+
+            rb = QRadioButton(text=widget.Model.name)
+            button_group.addButton(rb, i)
+            method_layout.addWidget(rb)
+            method_layout.addWidget(widget)
+
+        button_group.button(self.method_index).setChecked(True)
+        self.toggle_widgets()
+        method_layout.addStretch()
+
+        # Topics description
+        self.topic_desc = TopicViewer()
+        self.topic_desc.topicSelected.connect(self.send_topic_by_id)
+        self.mainArea.layout().addWidget(self.topic_desc)
+        self.topic_desc.setFocus()
+
+    def set_data(self, data=None):
+        self.corpus = data
+        self.apply()
+
+    def commit(self):
+        if self.corpus is not None:
+            self.apply()
+
+    @property
+    def model(self):
+        return self.widgets[self.method_index].model
+
+    def change_method(self, new_index):
+        if self.method_index != new_index:
+            self.method_index = new_index
+            self.toggle_widgets()
+            self.commit()
+
+    def toggle_widgets(self):
+        for i, widget in enumerate(self.widgets):
+            widget.setVisible(i == self.method_index)
+
+    def update_topics(self):
+        self.topic_desc.show_model(self.model)
+
+    @QtCore.pyqtSlot(int)
+    def progress(self, p):
+        self.progressBarSet(p, processEvents=None)
+
+    def apply(self):
+        self.stop_learning()
+        if self.corpus:
+            self.start_learning()
+        else:
+            self.send(Output.CORPUS, None)
+            self.send(Output.TOPIC, None)
+
+    def start_learning(self):
+        self.topic_desc.clear()
+        if self.corpus:
+            self.progressBarInit()
+
+            def progress_callback(i):
+                QMetaObject.invokeMethod(self, "progress", Qt.QueuedConnection, Q_ARG(int, i))
+
+            self.learning_thread = LearningThread(self.model, self.corpus.copy(),
+                                                  result_callback=self.send_corpus,
+                                                  progress_callback=progress_callback)
+            self.learning_thread.finished.connect(self.learning_finished)
+            self.learning_thread.start()
+
+    @QtCore.pyqtSlot()
+    def stop_learning(self):
+
+        if self.learning_thread and self.learning_thread.isRunning():
+            self.learning_thread.terminate()
+            self.learning_thread.wait()
+            self.progressBarFinished()
+
+    def send_corpus(self, corpus):
+        self.send(Output.CORPUS, corpus)
+        self.send_topic_by_id(0)
+
+    def learning_finished(self):
+        self.update_topics()
+        self.progressBarFinished()
+
+    def send_report(self):
+        self.report_items(*self.widgets[self.method_index].report_model())
+        if self.corpus:
+            self.report_items('Topics', [(i, self.model.get_top_words_by_id(i))
```

Comment:
I know I'm nitpicking here but I would prefer topics to have the same indices as in the widget (i.e. 1-based). Also words should be joined by commas so brackets and quotation marks are not seen in the report and that the same format is used as in the widget. I.e. something along those lines:

```py
self.report_items('Topics', [(i+1, ', '.join(self.model.get_top_words_by_id(i)))
                                        for i in range(len(self.model.topic_names))])
```